### ISSUE TITLE
Elasticsearch disk space

### ISSUE THREAD

**Issue by teddyking** [logsearch/logsearch-boshrelease#72]
Hi there,

We (CloudCredo) have recently run into a couple of issues with logsearch regarding elasticsearch and disk space.
I can see from [Issue #36](https://github.com/logsearch/logsearch-boshrelease/issues/36) that you have had some similar problems in the past.

Is there a recommended approach to dealing with this? I've been thinking about extending our logsearch deployment(s) with [curator](https://github.com/elasticsearch/curator) but I wanted to check if you had any plans to do something similar in the future?

Thanks,
Ed

**Comment by dpb587** [comment]
Hi Ed - thanks for taking the time to post! The disk space has been a common issue, and so far we've been using curator, too. Currently, our approach has been non-integrated - scheduling the curator task separately through a stable machine's cron or Jenkins.

But, we'd eventually like to see that cleanup/curator process integrated with the deployment. These are some of the implementation ideas we've internally touched on in the past, and I'd be very interested if you had other ideas or even wanted to give them a go yourself:
- curator as a BOSH errand (integrated, but doesn't inherently solve the scheduling aspect)
- a Jenkins build template which could be deployed somewhere in the network (non-integrated)
- curator as a job template with a long-running wrapper and lightweight lock (needs python dependency, probably installed on master nodes, the process would be sleeping 99% of the time between runs, use `_ttl` and elasticsearch docs for only running once if installed on multiple jobs)

So far, this issue hasn't been a priority amongst our other tasks. If you're able to make some progress on this that you can share, we'd love to see a PR. Or if you just want to chat implementation ideas, this issue will be a great place to do that, too.

**Comment by dpb587** [comment]
Another monit-based idea which only curates when space is limited...

```
check filesystem xvdfsomething with path /var/vcap/store
    if space usage > 90% then exec '/path/to/curator -d <%= p('elasticsearch.curator.retain_days') %> --other-options...
```

Wouldn't handle daily calls for index optimization though.

**Comment by mrdavidlaing** [comment]
I like the above idea which links curating to the actual problem trigger - lack of disk space

Monit also seems to have a CRON like scheduler, so we could also do a daily optimise like this:

```
check program elasticsearch_optimiser with
          path /var/vcap/jobs/elasticsearch/bin/optimise 
              timeout 3600 seconds
   every "* 2 * * *"
```

Since monit allows multiple check statements in the monit config for a specific job, we could have a collection of these.

**Comment by teddyking** [comment]
Hi guys thanks for the feedback!

I was originally hoping to run a curator command as a bosh errand on a recurring basis, but it doesn't look like bosh has the ability to schedule errands to run periodically (yet...).

I also like the look of the monit approach. A nice benefit I can see to this approach is that you wouldn't need to deploy any additional VMs, we could presumably run it as part of the existing jobs.

**Comment by mrdavidlaing** [comment]
It looks like we're converging on having a curator package embedded inside logsearch-boshrelease.

If I understand bosh errands correctly, this would be the first step towards implemented a curator-errand.

If we choose to go with monit triggering the curator run, a curator package is also required.

So, next step, make a curator package?

**Comment by teddyking** [comment]
Yep I'd agree with that. A curator package is going to be needed however we choose to run it. I'll let you know if/when I make any progress with this.

**Comment by dpb587** [comment]
@bonzofenix, @malston - fyi, here is where we were discussing curator and purging old indices.

**Comment by bonzofenix** [comment]
@dpb587 just started working on this. We will go ahead and implement it as a job and try to use the cron feature of monit.

**Comment by dpb587** [comment]
Great - keep us posted

**Comment by teddyking** [comment]
It looks like I actually started working on this a while ago: https://github.com/teddyking/logsearch-boshrelease

I can't remember how far I got, and it's probably out of date by now but I thought I'd link it here for reference.

**Comment by mrdavidlaing** [comment]
@teddyking - awesome - looks like you solved the difficult problem of all the curator python dependancies - https://github.com/teddyking/logsearch-boshrelease/commit/787c20e593ec906def658d032c5b490917530f30

I like the _way_ you have configured curator - https://github.com/teddyking/logsearch-boshrelease/commit/003d1760945c2c20d30502efa19b964fd14d6147 - but I don't think it should be _part_ of the `elasticsearch` job; since then you'd have multiple instances all running at once on larger logsearch installations.

@dpb587, @bonzofenix - Do you think curator should become part of the new `elasticsearch_config` job, or be separate?

**Comment by mrdavidlaing** [comment]
Hmm.  Just noticed why @teddyking co-located with the `elasticsearch` job; so the `curator` run can be triggered by monit detecting low disk space.  

For the first version of this feature, I think we should just stick with the simpler `curator` CRON approach that does a nightly purge of all indices older than `retention_period` days.

**Comment by bonzofenix** [comment]
@mrdavidlaing @teddyking @dpb587 -  We got as far as getting curator package with its dependencies installed with pip and a python3 package. 
Unfortunately we run into a problem ([see thread](https://lists.cloudfoundry.org/archives/list/cf-bosh@lists.cloudfoundry.org/thread/AXSGBURLS5TK4SG43G5PWUJTDO2XSGPW/)) . 
Till we solve this we think the best approach, is to make it as an errand and wait for comments on this from the community. We do not think it worth going with a monit process type approach. 

Do you guys agree on this? 
when we get an elegant option to do the cron part we could go for it.

@bonzofenix && @grayisgreat

**Comment by dpb587** [comment]
I'm okay with the errand type approach for now - you've taken care of the hard part of getting everything packaged up nicely, so we can refactor the scheduling side of things based on the list recommendations later. If you're in a mood to automate though, a couple alternatives to consider...

**Utilize the host's crond.** We use this with the [archiver job](https://github.com/logsearch/logsearch-boshrelease/blob/develop/jobs/archiver/templates/bin/logstash.process.erb#L48-L59) in order to ship off files at a configurable schedule. This approach has the drawback of not forwarding a failure exit code if something goes wrong, but assuming you have some basic alarms on disk space usage or the stderr logs it should be quickly noticed.

**Infinite loop.** Stick the job's curator call inside an infinite loop and have a configurable `sleep 86400` after every call. Also hacky.

**Use monit's process uptime testing.** Haven't tried this one, but I think I recall BOSH's monit version supports [uptime](https://mmonit.com/monit/documentation/monit.html#PROCESS-UPTIME-TESTING) checks. During startup it could run curator and pretend to be PID 1. After monit thinks its been up more than 24 hours, restart wherein stop simply deletes its PID file of 1 and start kicks things off again. A bit hacky, though.

**Comment by mrdavidlaing** [comment]
@bonzofenix, @grayisgreat, @dpb587,  After consideration, I'm against shipping this as an errand.

I think its better to go with using host's crond, as per the archiver job, and then upgrade that to something better if / when that materializes.  

Otherwise we force users to figure out a way to schedule running of an errand (a non trivial task); and then have them throw that work away once we move the scheduling into the job.

**Comment by bonzofenix** [comment]
@mrdavidlaing @dpb587 @teddyking We went ahead and use the crond on the host machine. it is working, we are doing some code refactoring and the pull request should be there tomorrow.

**Comment by mrdavidlaing** [comment]
@bonzofenix :guitar: on!

**Comment by dpb587** [comment]
Thanks everyone for the discussion and time on this. We've merged in a curator cron job via PR #170 (from @bonzofenix, @cloud-dude, et al) which will be included in the next logsearch release.
